packages/flutter_tools/lib/src/test/coverage_collector.dart

```diff
@@ -187,8 +187,13 @@ class CoverageCollector extends TestWatcher {
     if (formatter == null) {
       final coverage.Resolver usedResolver = resolver ?? this.resolver ?? await CoverageCollector.getResolver(packagesPath);
       final String packagePath = globals.fs.currentDirectory.path;
-      final List<String> reportOn = coverageDirectory == null
-          ? <String>[globals.fs.path.join(packagePath, 'lib')]
+      // find paths for libraryNames so we can include them to report
+      final List<String>? libraryPaths = libraryNames
+          ?.map((String e) => usedResolver.resolve('package:$e'))
+          .whereType<String>()
+          .toList();
+      final List<String>? reportOn = coverageDirectory == null
```

Comment:
Won't this behavior be surprising to the user if they've provided `--package-name-regex` but the `coverageDirectory` is non-null, and their regexes are ignored?

Reply:
Let's first take a look at it in scope of the `CoverageCollector` class. This class doesn't know about command line options that we provide. It accepts `libraryNames` in the constructor, which is 
> The names of the libraries to gather coverage for. If null, all libraries will be accepted.

So by providing multiple library names, I would expect this class to collect coverage for them (in the `collectCoverage` function, existing behavior), and I would expect this collected data to be included in the report (`finalizeCoverage` function, surprisingly **not** the existing behavior, but it is added in this PR, in the lines we currently discuss).  
Now, if we want to filter some files out (by path) from the final report, we can provide `coverageDirectory` argument to the `finalizeCoverage`. Here we've already collected coverage data, but use `coverageDirectory` to filter some of it from the report. Therefore `libraryNames` are not ignored and libraries will be included in the report if they are inside `coverageDirectory` (if provided).

--- 

In scope of the flutter `test` command, `coverageDirectory` is simply not provided to `finalizeCoverage`:
https://github.com/flutter/flutter/blob/ce9a604a3bad2c56067d8fac2a5251d17c312fbf/packages/flutter_tools/lib/src/commands/test.dart#L492-L495
https://github.com/flutter/flutter/blob/ce9a604a3bad2c56067d8fac2a5251d17c312fbf/packages/flutter_tools/lib/src/test/coverage_collector.dart#L201-L204

But even if it will, it won't break the logic of the `finalizeCoverage` function.

Reply:
ok, it looks like these code paths are mutually exclusive.